### PR TITLE
update UITableView and UICollectionView Extensions

### PR DESCRIPTION
**Pull request by jdisho**
In this PR I am updating:
`dequeueReusableCell(withClass:for)`,
`dequeueReusableHeaderFooterView(withClass)`,
`dequeueReusableCell(withClass)` in **UITableViewExtentions** and  `dequeueReusableCell(withClass:for)`,
`dequeueReusableSupplementaryView(ofKind:withClass:for)` in **UICollectionViewExtentions** 
to return not an optional value. In my opinion is better to get a `fatalError(...)` if the nib file could not be found.

## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 4.
- [x] New extensions support iOS 8.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

File: Sources/Extensions/UIKit/UICollectionViewExtensions.swift [hunk 1]
```diff
@@ -75,9 +75,12 @@ public extension UICollectionView {
 	///   - name: UICollectionViewCell type.
 	///   - indexPath: location of cell in collectionView.
 	/// - Returns: UICollectionViewCell object with associated class name.
-	public func dequeueReusableCell<T: UICollectionViewCell>(withClass name: T.Type, for indexPath: IndexPath) -> T? {
-		return dequeueReusableCell(withReuseIdentifier: String(describing: name), for: indexPath) as? T
-	}
+    public func dequeueReusableCell<T: UICollectionViewCell>(withClass name: T.Type, for indexPath: IndexPath) -> T {
+        guard let cell = self.dequeueReusableCell(withReuseIdentifier: String(describing: name), for: indexPath) as? T else {
+            fatalError("Couldn't find nib file for \(String(describing: name))")
```
Comment (SD10) [inline Sources/Extensions/UIKit/UICollectionViewExtensions.swift]: This one is not a nib 🤔

File: Sources/Extensions/UIKit/UICollectionViewExtensions.swift [hunk 2]
```diff
@@ -86,8 +89,11 @@ public extension UICollectionView {
 	///   - name: UICollectionReusableView type.
 	///   - indexPath: location of cell in collectionView.
 	/// - Returns: UICollectionReusableView object with associated class name.
-	public func dequeueReusableSupplementaryView<T: UICollectionReusableView>(ofKind kind: String, withClass name: T.Type, for indexPath: IndexPath) -> T? {
-		return dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: String(describing: name), for: indexPath) as? T
+	public func dequeueReusableSupplementaryView<T: UICollectionReusableView>(ofKind kind: String, withClass name: T.Type, for indexPath: IndexPath) -> T {
+        guard let cell = dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: String(describing: name), for: indexPath) as? T else {
+            fatalError("Couldn't find nib file for \(String(describing: name))")
```
Comment (SD10) [inline Sources/Extensions/UIKit/UICollectionViewExtensions.swift]: Same here

**Comment by LucianoPAlmeida** [comment]
Hey @jdisho @SD10 I agree that is reasonable, but point to be discussed point is that instead of a fatalError() use precondition() or assert()? 
Another point maybe it's if have a fatal error inside the extension is a good idea, I don't know maybe is something to let the end user do on his own implementation. 
Let me know what you guys think :))

**Comment by SD10** [comment]
@LucianoPAlmeida Hmmmm. I'm using `fatalError()` in all my projects for this scenario. We use `fatalError()` in both MessageKit and GitHawk

**Comment by LucianoPAlmeida** [comment]
@SD10 It's kinda personal I think some people prefer use precondition, some use fatal error.
 I don't think fatal error is a problem 😂just want to point it out :))

**Comment by jdisho** [comment]
@LucianoPAlmeida I agree that is a personal preference to choose between those, but IMO `fatalError` implementation in the extension fits best to this use case 😊 .
@SD10 I added a changelog entry about my changes 😊

**Comment by SD10** [comment]
Thank you for contributing to SwifterSwift! I've invited you to join the SwifterSwift GitHub organization - no pressure to accept! If you'd like more information on what that means, check out our [contributing guidelines](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md). Feel free to reach out if you have any questions! 😃

**Pull request merged**